### DISCUSSION TITLE
Feedback on the Merge Queue (Beta)

### DISCUSSION THREAD

**Discussion by Gallimathias** [community/community#14333] [Pull Requests]
I haven't found another place for the feedback, so I think this is the right place. Otherwise, please feel free to move it.

We got access in the company in March. The merge queue is a damn good thing and solves some problems we have actively.    

Currently we had to turn the queue off again for the following reasons:

* State not clear enough: 
  * If a build fails on the queue you can't tell what failed by looking at the PR. It takes a lot of research to find out why a PR did not finally reach its goal. This is what finally stopped our test series. Everything built in the queue, but issues and success it just wasn't noticed in the end.
  * The builds that are currently running on the queue are not accessible from the PR, which makes it impossible for the PR creator and the reviewer to see whether something is happening on the queue and what is happening.
* Automerging allows a loophole:
  * There is no possibility to individually tell the system that you want to have different number of reviewers required per PR, you can just say for the hole project a number of reqired reviews. Since the merge queue also activates the automerge, it is now possible, even though you have requested two (required) reviewers for a specific case (normaly we need only one reviewer but on some special cases more), that the automerge queued the pr directly after the first pr for merging. It would be nice if we could limited this in some special cases more.
* Extended build time:
  * The merge queue has led to more build times and thus to longer waiting times. Although several PR's were already up to date and theoretically could have been merged directly, they were built again in the queue. This ultimately led to us having to wait an additional 30 minutes per PR instead directly merge and final build the PR for the final artifacts.

I hope the feedback helps you to develop the feature further.

**Comment by antonagestam** [comment]
> Extended build time

This should be somewhat expected from a merge queue, because more work will happen _before_ a commit lands on trunk. With a merge queue you will have the guarantee that every commit that lands on your trunk branch has already passed CI (the exact commit that lands, not a stale commit on the feature branch).

This means that you should be able to reduce the amount of work that happens _after_ the commit has landed on your trunk branch, e.g. it's unnecessary to trigger CI to run on `push` on your `main` branch, that will just be double work.

One could argue that Github should notice when a branch is really in a "fast-forwardable state" and skip running CI if it has already run on the PR. But, when used in conjunction with "squash on merge", that would violate the above mentioned guarantee that every commit that lands on trunk has already passed CI, as its not that exact commit that would have passed.

**Comment by brianhenn** [comment]
Similarly we tested the merge queue feature (February 2023) and dropped it. For us, the main missing functionality was that the enforced squash merge did not allow for modifying the squash commit text, which is something we usually do to make the commit history more informative and align with the information in the PR descriptions. Additionally, we ran into the [issue](https://github.com/orgs/community/discussions/38199) (solvable, I think but more time consuming than it should be) of CircleCI not integrating well with the GH merge queue.

**Comment by smailc** [comment]
Does reducing the build concurrency right down just pause the PRs at the bottom of the queue from building? I presume it doesn't cause them to fail?
